### DISCUSSION TITLE
php artisan schedule:run add runInBackground wait

### DISCUSSION THREAD

**Discussion by seth-shi** [laravel/framework#51418] [General]
* tasks aaa need five minutes run

```
 protected function schedule(Schedule $schedule): void
    {
        $schedule->command('aaa')->cron('* * * * *')->runInBackground();
      $schedule->command('bbb')->cron('* * * * *')->runInBackground();
       $schedule->command('ccc')->cron('* * * * *')->runInBackground();
    }
```

* i use k8s cronjob run `php artisan schedule:run`, but `php artisan schedule:run` The task was quickly completed, as he put aaa running in the background
* then pod quit, background run `aaa` quit

maybe can add paramters `php artisan schedule:run --wait`

**Comment by ghost** [comment]
This is a problem from your infra. The container running the scheduler must be up always.

**Reply by seth-shi** [reply]
you need try k8s cronjob

**Reply by ghost** [reply]
I don't use that:) I use ECS + task definition and the scheduler has its own container. I don't have this issue. Your issue is not in laravel.

The scheduler starts in BG all processes, then exists. All BG processes will take their time to finish.

I suggest asking the question ALSO in K8S specialized forum, maybe you will get a solution.

PS. FYI logging to stdout will not work from the BG (shell) processes (https://github.com/laravel/framework/discussions/50085).

**Reply by seth-shi** [reply]
if you use docker, once container execute command is normal;

```
docker exec --rm laravel php artisan schedule:run
```

**Reply by ghost** [reply]
K8S will spin up a container just to run shedule:run ?

"--rm removes the stopped container (the one docker run creates), not the image it was based on (what you downloaded). "(https://stackoverflow.com/questions/49726272/what-is-the-rm-flag-doing)

**Reply by seth-shi** [reply]
k8s use cronjob, every one minute execute `php artisan schedule:run`
docker use `docker exec --rm laravel php artisan schedule:run` for every minute execute

why would I do that, because code update, i can update container image, then `docker exec --rm laravel php artisan schedule:run`  use latest image run

**Reply by ghost** [reply]
remove the --rm and try again

when you change the image you restart the container manually from the new image

**Reply by seth-shi** [reply]
```
 protected function schedule(Schedule $schedule): void
    {
        $schedule->command('aaa')->cron('* * * * *')->runInBackground();
      $schedule->command('bbb')->cron('* * * * *')->runInBackground();
       $schedule->command('ccc')->cron('* * * * *')->runInBackground();
    }
```
if has aaa cronjob need run ten minutes, It ran for five minutes, i need update code, do i restart container? that aaa cronjob is fail

**Reply by ghost** [reply]
![image](https://github.com/laravel/framework/assets/82870454/d99df490-3403-43dc-ba14-081ad2c65710)
If the container is up then set time limit in the job if it needs more time than default from php ini.
see https://www.php.net/manual/en/function.set-time-limit.php

**Reply by seth-shi** [reply]
can you give me a answer, if container is aiways running, so how i update container by new image and old job don't stop

**Reply by ghost** [reply]
From a devops perspective no. 
From a logical point of view, you must not kill the old containers until all processes had finished and not start new ones.

**Reply by seth-shi** [reply]
How do you know if the process is complete?

**Reply by ghost** [reply]
For one queue worker I found this solution: https://github.com/laravel/framework/discussions/49954
Maybe you can tweak it for scheduler also.

**Reply by seth-shi** [reply]
i don't need queue, i need schedule. 
for queue i used `php artisan queue:work`, then i  update container by new image can send SIGQUIT, `php artisan queue:work` can exit, then i update container

**Reply by macropay-solutions** [reply]
@seth-shi We developed a solution that ensures that MULTIPLE queue workers and MULTIPLE background cron processes started by the scheduler are not killed before exiting, during a deploy or scale-down. If you are a company, please contact us.
